vortex-array/src/patches.rs

```diff
@@ -248,10 +355,63 @@ impl Patches {
                     .search_sorted(&needle, SearchSortedSide::Left);
             });
         }
-        self.indices.as_primitive_typed().search_sorted(
-            &PValue::U64((index + self.offset) as u64),
-            SearchSortedSide::Left,
-        )
+        indices
+            .as_primitive_typed()
+            .search_sorted(&PValue::U64(needle as u64), SearchSortedSide::Left)
+    }
+
+    /// Constant time searches for `index` in the indices array.
+    ///
+    /// First determines which chunk the target index falls into, then performs
+    /// a binary search within that chunk's range.
+    ///
+    /// Returns a [`SearchResult`] indicating either the exact patch index if found,
+    /// or the insertion point if not found.
+    ///
+    /// # Panics
+    /// Panics if `chunk_offsets` or `offset_within_chunk` are not set.
+    fn search_index_chunked(&self, index: usize) -> SearchResult {
+        let Some(chunk_offsets) = &self.chunk_offsets else {
+            vortex_panic!("chunk_offsets is required to be set")
+        };
+
+        let Some(offset_within_chunk) = self.offset_within_chunk else {
+            vortex_panic!("offset_within_chunk is required to be set")
+        };
+
+        let chunk_start_idx = (index + self.offset % PATCH_CHUNK_SIZE) / PATCH_CHUNK_SIZE;
+        let chunk_end_idx = chunk_start_idx + 1;
+
+        // Patch index offsets are absolute and need to be offset by the first chunk of the current slice.
+        let base_offset = self.chunk_offset_at(0);
+
+        let patches_start_idx = (self.chunk_offset_at(chunk_start_idx) - base_offset)
+            // Chunk offsets are only sliced off in case the slice is fully
+            // outside of the chunk range.
+            //
+            // Though the range for indices and values is sliced in terms of
+            // individual elements, not chunks. To account for that we do a
+            // saturating sub when adjusting the indices based on the chunk offset.
+            .saturating_sub(offset_within_chunk);
+
+        let patches_end_idx = if chunk_end_idx < chunk_offsets.len() {
```

Comment:
I think we can slightly rewrite this since the notion that we have chunk start and end is a bit misleading 
```suggestion
        let patches_end_idx = if chunk_start_idx < chunk_offsets.len() - 1 {
```
and rename chunk_start_idx to chunk_idx

Reply:
done